src/transformers/models/fnet/modeling_fnet.py

```diff
@@ -0,0 +1,1063 @@
+        super().__init__()
+        self.fourier_transform = torch.fft.fftn
+
+    def forward(
+        self,
+        hidden_states,
+    ):
+        return (self.fourier_transform(hidden_states).real,)
+
+
+class FNetBasicOutput(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        # TODO: Check if removal of a dense layer and dropout is okay.
+        self.LayerNorm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+
+    def forward(self, hidden_states, input_tensor):
+        hidden_states = self.LayerNorm(input_tensor + hidden_states)  # NOTE: Order has been changed.
+        return hidden_states
+
+
+class FNetFourierTransform(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.self = FNetBasicFourierTransform(config)
+        self.output = FNetBasicOutput(config)
+
+    def forward(
+        self,
+        hidden_states,
+    ):
+        self_outputs = self.self(
+            hidden_states,
+        )
+        fourier_output = self.output(self_outputs[0], hidden_states)
+        outputs = (fourier_output,)
+        return outputs
+
+
+class FNetIntermediate(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.dense = nn.Linear(config.hidden_size, config.intermediate_size)
+        if isinstance(config.hidden_act, str):
+            self.intermediate_act_fn = ACT2FN[config.hidden_act]
+        else:
+            self.intermediate_act_fn = config.hidden_act
+
+    def forward(self, hidden_states):
+        hidden_states = self.dense(hidden_states)
+        hidden_states = self.intermediate_act_fn(hidden_states)
+        return hidden_states
+
+
+class FNetOutput(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.dense = nn.Linear(config.intermediate_size, config.hidden_size)
+        self.LayerNorm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+
+    def forward(self, hidden_states, input_tensor):
+        hidden_states = self.dense(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+        hidden_states = self.LayerNorm(input_tensor + hidden_states)
+        return hidden_states
+
+
+class FNetLayer(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        # TODO: Check if chunk_size_feed_forward is needed
+        self.chunk_size_feed_forward = config.chunk_size_feed_forward
+        self.seq_len_dim = 1
+        self.fourier = FNetFourierTransform(config)
+        self.intermediate = FNetIntermediate(config)
+        self.output = FNetOutput(config)
+
+    def forward(
+        self,
+        hidden_states,
+    ):
+
+        self_fourier_outputs = self.fourier(
+            hidden_states,
+        )
+        fourier_output = self_fourier_outputs[0]
+
+        # # TODO: Check if this should be here.
+        layer_output = apply_chunking_to_forward(
```

Comment:
If it works - great! Let's leave it :-) The purpose of the feed-forward chunking is explained here: https://huggingface.co/blog/reformer#2-chunked-feed-forward-layers